File: ClimaDaemon/Core/Clima.Core.Scheduler/Network/Services/SchedulerControlService.cs
using Clima.Basics.Services.Communication;
using Clima.Core.DataModel;
using Clima.Core.Network.Messages;
using Clima.Core.Scheduler.Network.Messages;

namespace Clima.Core.Scheduler.Network.Services
{
    public class SchedulerControlService:INetworkService
    {
        private readonly IClimaScheduler _scheduler;

        public SchedulerControlService(IClimaScheduler scheduler)
        {
            _scheduler = scheduler;
        }

        [ServiceMethod]
        public DefaultResponse SetTemperatureProfile(SetProfileRequest request)
        {
            _scheduler.SetTemperatureProfile(request.ProfileKey);
            return new DefaultResponse();
        }

        [ServiceMethod]
        public SchedulerProcessInfo GetProcessInfo(DefaultRequest request)
        {
            return _scheduler.SchedulerProcessInfo;
        }

        [ServiceMethod]
        public SchedulerProfilesInfo GetProfilesInfo(DefaultRequest request)
        {
            return _scheduler.SchedulerProfilesInfo;
        }
        
        [ServiceMethod]
        public DefaultResponse SetVentilationProfile(SetProfileRequest request)
        {
            _scheduler.SetVentilationProfile(request.ProfileKey);
            return new DefaultResponse();
        }
        [ServiceMethod]
        public DefaultResponse SetValveProfile(SetProfileRequest request)
        {
            _scheduler.SetValveProfile(request.ProfileKey);
            return new DefaultResponse();
        }
        [ServiceMethod]
        public DefaultResponse SetMineProfile(SetProfileRequest request)
        {
            _scheduler.SetMineProfile(request.ProfileKey);
            return new DefaultResponse();
        }

        [ServiceMethod]
        public SchedulerDebugResponse GetSchedulerDebugInfo(DefaultRequest request)
        {
            return new SchedulerDebugResponse();
        }

        [ServiceMethod]
        public VentilationParamsResponse UpdateVentParameters(VentilationParamsRequest request)
        {
            _scheduler.VentilationParameters = request.Parameters;
            return new VentilationParamsResponse()
            {
                Parameters = _scheduler.VentilationParameters
            };
        }

        [ServiceMethod]
        public VentilationParamsResponse GetVentilationParams(DefaultRequest request)
        {
            return new VentilationParamsResponse() {Parameters = _scheduler.VentilationParameters};
        }
    }
}

File: QtClient/QtClimaClient/Network/GenericServices/SchedulerControlService.cpp
#include "SchedulerControlService.h"

#include <Network/GenericServices/Messages/SchedulerInfoResponse.h>
#include <Network/GenericServices/Messages/SetProfileRequest.h>
#include <Network/GenericServices/Messages/VentilationParamsRequest.h>

SchedulerControlService::SchedulerControlService(QObject *parent) : INetworkService(parent)
{

}

void SchedulerControlService::SetTemperatureProfile(const QString &profileKey)
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "SchedulerControlService";
    request->method = "SetTemperatureProfile";
    SetProfileRequest rq;
    rq.ProfileKey = profileKey;

    request->params = rq.toJsonString();
    emit SendRequest(request);
}

void SchedulerControlService::SetVentilationProfile(const QString &profileKey)
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "SchedulerControlService";
    request->method = "SetVentilationProfile";
    SetProfileRequest rq;
    rq.ProfileKey = profileKey;

    request->params = rq.toJsonString();
    emit SendRequest(request);
}

void SchedulerControlService::SetValveProfile(const QString &profileKey)
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "SchedulerControlService";
    request->method = "SetValveProfile";
    SetProfileRequest rq;
    rq.ProfileKey = profileKey;

    request->params = rq.toJsonString();
    emit SendRequest(request);
}

void SchedulerControlService::SetMineProfile(const QString &profileKey)
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "SchedulerControlService";
    request->method = "SetMineProfile";
    SetProfileRequest rq;
    rq.ProfileKey = profileKey;

    request->params = rq.toJsonString();
    emit SendRequest(request);
}

void SchedulerControlService::GetProfilesInfo()
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "SchedulerControlService";
    request->method = "GetProfilesInfo";
    emit SendRequest(request);
}

void SchedulerControlService::GetProcessInfo()
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "SchedulerControlService";
    request->method = "GetProcessInfo";
    emit SendRequest(request);
}

void SchedulerControlService::GetVentilationParams()
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "SchedulerControlService";
    request->method = "GetVentilationParams";
    emit SendRequest(request);
}

void SchedulerControlService::UpdateVentilationParams(VentilationParams parameters)
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "SchedulerControlService";
    request->method = "UpdateVentParameters";

    VentilationParamsRequest pr;
    pr.Parameters = parameters;

    request->params = pr.toJsonString();
    emit SendRequest(request);
}

void SchedulerControlService::ProcessReply(NetworkResponse *reply)
{
    if(reply->service == "SchedulerControlService")
    {
        if(reply->method == "GetProfilesInfo")
        {
            SchedulerProfilesInfo info;
            info.fromJson(reply->result.toUtf8());
            emit SchedulerProfilesInfoReceived(info);
        }
        else if(reply->method == "GetProcessInfo")
        {
            SchedulerProcessInfo info;
            info.fromJson(reply->result.toUtf8());
            emit SchedulerProcessInfoReceived(info);
        }
        else if(reply->method == "SetTemperatureProfile" ||
                reply->method == "SetVentilationProfile" ||
                reply->method == "SetValveProfile" ||
                reply->method == "SetMineProfile")
        {
            emit SchedulerUpdated();
        }
        else if(reply->method == "GetVentilationParams")
        {
            VentilationParams params;
            params.fromJson(reply->result.toUtf8());
            emit VentilationParamsReceived(params);
        }
        else if(reply->method == "UpdateVentilationParams")
        {
            VentilationParams params;
            params.fromJson(reply->result.toUtf8());
            emit VentilationParamsUpdated(params);
        }
    }
}
